design-documents/service-isolation/detailed-architecture.md

```diff
@@ -0,0 +1,87 @@
+# Detailed architecture of multi-service Magento
+
+## Overview
+
+The goal of this document is to gather requirements, understand data flow and propose architecture for multi-service Magento application.
+
+Please make sure to get familiar with [high-level service isolation vision](../service-isolation.md) before proceeding.
+
+![Gateway](img/GraphQL%20BFF.png)
+
+The client (e.g. browser) is interacting with the system using GraphQL. The only component of the system exposed publicly is Storefront Gateway, which is responsible for routing, full response caching and perimeter-level authorization.
+
+Storefront Gateway routes incoming queries between PWA, GraphQL Engine and Storefront Authentication Gateway.
+
+PWA's main responsibility is to [render React application on the server-side](../frontend/server-side-rendering.md). Resulting HTML page with React application becomes cacheable on the server.
+
+GraphQL server parses GraphQL query, enriches the request (explained later) and proxies calls to the Storefront services. It also manages Customer context JWT. GraphQL schema is defined here and is extensible, query resolution is based on generic resolvers and configuration for field mapping to Storefront APIs.
+
+Storefront services are responsible for authorizing requests based on the Auth JWT and the list of required permissions. Actual ACL check is delegated to the the Authorization service.
+
+GraphQL authentication gateway is an independent service for improved security purposes, although it can be merged with GraphQL server.
+
+The storefront services are communicating with each other using REST or Protobuf protocol. Storefront APIs are implemented as service contracts exposed via REST/Protobuf.
+
+## Authentication, Store and Customer contexts
+
+**Auth JWT** - authentication token, which contains encrypted customer ID and is used on GraphQL Server to authenticate the logged-in customers.
```

Comment:
I think we should pass token using cookie with HttpOnly flag (this is how PHP session cookie being passed) instead of a header. But in this case, why wouldn't we use session?

Reply:
What about other applications like mobile which don't use cookies?

Reply:
For other applications it might be ok to use token. Problem with storing token in web applications is that it can easily stollen.

Reply:
It can be stored in the same place as cookie, please give some references which explain risks and better solutions.